frontend/gateway/gateway.go

```diff
@@ -738,6 +744,442 @@ func (lbf *llbBridgeForwarder) Inputs(ctx context.Context, in *pb.InputsRequest)
+
+		pioDone := make(chan string)
+		defer close(pioDone)
+
+		for {
+			var execMsg *pb.ExecMessage
+			select {
+			case pid := <-pioDone:
+				pio := pios[pid]
+				delete(pios, pid)
+				if pio != nil {
+					err := pio.Close()
+					if err != nil {
+						return err
+					}
+				}
+			case <-ctx.Done():
+				return ctx.Err()
+			case execMsg = <-msgs:
+			}
+
+			if execMsg == nil {
+				continue
+			}
+
+			pid := execMsg.ProcessID
+			if pid == "" {
+				return status.Errorf(codes.InvalidArgument, "ProcessID required")
+			}
+
+			pio, pioFound := pios[pid]
+
+			if data := execMsg.GetFile(); data != nil {
+				if !pioFound {
+					return status.Errorf(codes.NotFound, "IO for process %q not found", pid)
+				}
+				err := pio.Write(data)
+				if err != nil {
+					return err
+				}
+			} else if resize := execMsg.GetResize(); resize != nil {
+				if !pioFound {
+					return status.Errorf(codes.NotFound, "IO for process %q not found", pid)
+				}
+				pio.Resize(resize)
+			} else if init := execMsg.GetInit(); init != nil {
+				if pioFound {
+					return status.Errorf(codes.AlreadyExists, "Process %s already exists", pid)
+				}
+				id := init.ContainerID
+				lbf.ctrsMu.Lock()
+				ctr, ok := lbf.ctrs[id]
+				lbf.ctrsMu.Unlock()
+				if !ok {
+					return status.Errorf(codes.NotFound, "container %q previously released or not created", id)
+				}
+
+				initCtx, initCancel := context.WithCancel(ctx)
+				defer initCancel()
+
+				pio := newProcessIO(initCtx.Done(), init.Fds)
+				pios[pid] = pio
+
+				// on cancel or if container is released then
+				// mark pio as done so it will be cleaned up
+				eg.Go(func() error {
+					select {
+					case <-initCtx.Done():
+					case <-ctr.release:
+						initCancel()
+					}
+					pioDone <- pid
+					return nil
+				})
+
+				process := &executor.ProcessInfo{
+					Meta: executor.Meta{
+						Args: init.Meta.Args,
+						Env:  init.Meta.Env,
+						User: init.Meta.User,
+						Cwd:  init.Meta.Cwd,
+						Tty:  init.Tty,
+					},
+					Stdin:  pio.fds[0],
+					Stdout: pio.fds[1],
+					Stderr: pio.fds[2],
+					Resize: pio.resize,
+				}
+				process.Meta.Env = addDefaultEnvvar(process.Meta.Env, "PATH", utilsystem.DefaultPathEnv)
+				if init.Tty {
+					process.Meta.Env = addDefaultEnvvar(process.Meta.Env, "TERM", "xterm")
+				}
+
+				for fd, file := range pio.fds {
+					file := file
+					// don't read from container stdin
+					if fd == 0 || file == nil {
+						continue
+					}
+					eg.Go(func() error {
+						dest := &outputWriter{
+							stream:      srv,
+							fd:          uint32(fd),
+							processID:   pid,
+							containerID: id,
+						}
+						_, err := io.Copy(dest, file)
+						if err == nil {
+							// no error so must be EOF
+							err = srv.Send(&pb.ExecMessage{
+								ProcessID: pid,
+								Input: &pb.ExecMessage_File{
+									File: &pb.FdMessage{
+										Fd:  uint32(fd),
+										EOF: true,
+									},
+								},
+							})
+						}
+						return err
```

Comment:
Does this `io.Copy` not return `io.EOF`?

If it doesn't, I think this can be just arranged like:
```go
_, err := io.Copy(dest, file)
if err != nil {
	return err
}

// no error so must be EOF
return srv.Send(&pb.ExecMessage{
	ProcessID: pid,
	Input: &pb.ExecMessage_File{
		File: &pb.FdMessage{
			Fd:  uint32(fd),
			EOF: true,
		},
	},
})
```

Reply:
> A successful Copy returns err == nil, not err == EOF. Because Copy is defined to read from src until EOF, it does not treat an EOF from Read as an error to be reported.

So yeah, should report nil on EOF, I will fix as you suggest.